### PR TITLE
[Issue #2809] Handle parsing the jwt we created, and connect to a user

### LINKED ISSUE #2809: Create a new auth implementation that validates a JWT token and attaches a user to the global context
### Summary

TODO - some of the technical details.

We want a new auth approach that we can use as an alternative to our current:
`@opportunity_blueprint.auth_required(api_key_auth)`

Instead we should have a jwt auth that we can attach to our routes. This will do a few things:
* Fetch a token from a header param
* Validate that the token is one we generated in our API (see #2808 )
* Fetch a user from the DB with the given user ID
  * Attach this user object to the global context to be fetched elsewhere

Open questions:
* How does this work with the DB session management, we should be able to get a DB session in other methods, but will the user itself be funky because of a detached session issue? 

### Acceptance criteria

- [ ] Auth implementation created
- [ ] Thorough unit tests written to verify edge case scenarios

### PR DESCRIPTION
**Pull request by chouinar**
## Summary
Fixes #2809 

## Changes proposed
Setup logic to process the jwt we created in https://github.com/HHS/simpler-grants-gov/pull/2898

Setup a method to automatically generate a key for local development in a secure way via an override file

## Context for reviewers
The core part of this PR is pretty straightforward, we parse the JWT, do some validation, raise specific error messages for certain scenarios, and have tests for that behavior.

For the auth token in the API request header, instead of using `Bearer ..` I left it as a dedicated header field. The bearer format doesn't let you specify the header name and if we ever need multiple tokens supported in an endpoint will lead to more headache.

---

Where things got complex was setting up the private/public key for the API. These just need to be stored in env vars, but putting them directly in our local.env file isn't ideal - even though the key will be distinctly local-only, it will always be something flagged in security scans and just generally look problematic.

To work around this fun problem, I realized I could solve another annoyance at the same time, Docker as of January 2024 allows you to specify multiple env files + make them optional. So - I used that. I setup a script that creates an `override.env` file that you can freely modify and won't be checked in, and more importantly, automatically contains secrets like those public/private keys we didn't want to check in. (Note - if you're wondering why I didn't use Docker secrets, they're far more complex and this PR would've been 20+ files to make that half-work).

## Additional information
Locally I confirmed we can set tokens in the swagger docs and they work - we don't yet have an endpoint that uses this outside of the unit test I setup, but I temporarily modified the healthcheck endpoint to validate things work outside of tests as well.

Screenshot 2024-11-20 at 4 06 48 PM

The override file we generate looks like this (with the relevant key info removed):
```
# override.env
#
# Any environment variables written to this file
# will take precedence over those defined in local.env
#
# This file will not be checked into github and it is safe
# to store secrets here, however you should still follow caution
# with using any secrets locally if they cause the app to interact
# with external systems.
#
# This file was generated by running:
#    make setup-env-overrides
#
# Which runs as part of our "make init" flow.
#
# If you would like to re-generate this file, please run:
#    make setup-env-overrides --recreate
#
# Note that this will completely erase any existing configuration you may have


############################
# Authentication
############################

API_JWT_PRIVATE_KEY="-----BEGIN RSA PRIVATE KEY-----
...
-----END RSA PRIVATE KEY-----"

API_JWT_PUBLIC_KEY="-----BEGIN PUBLIC KEY-----
...
-----END PUBLIC KEY-----"
```

### CODE REVIEW

**Review by mdragon — APPROVED** [review]

File: api/bin/setup-env-override-file.sh [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+  PUBLIC_KEY=`cat tmp_jwk.pub`
+  PRIVATE_KEY=`cat tmp_jwk.key`
+
+  cat > $OVERRIDE_FILE <<EOF
+# override.env
+#
+# Any environment variables written to this file
+# will take precedence over those defined in local.env
+#
```
Comment (coilysiren) [inline api/bin/setup-env-override-file.sh:67]: fancy docs!

**Review by coilysiren — DISMISSED** [review]

File: api/docker-compose.yml [hunk 2]
```diff
@@ -90,7 +90,11 @@ services:
         "--reload",
       ]
     container_name: grants-api
-    env_file: ./local.env
+    env_file:
+      - path: ./local.env
+        required: true
+      - path: ./override.env
```
[the remark this answers is not in the document]
Comment (chouinar) [reply api/docker-compose.yml:96]: Yeah, when I came across it in the docs I realized it was exactly what I'd been looking for in env var management from Docker for years

**Review by coilysiren — APPROVED** [review]

**Review by mdragon — APPROVED** [review]
Looks good

**Review by coilysiren — APPROVED** [review]
